contrib/epee/include/net/connection_basic.hpp

```diff
@@ -84,15 +86,53 @@ class connection_basic { // not-templated base class for rapid developmet of som
     /// Strand to ensure the connection's handlers are not called concurrently.
     boost::asio::io_service::strand strand_;
     /// Socket for the connection.
-    boost::asio::ip::tcp::socket socket_;
+    ssl_context_t &m_ssl_context;
+    ssl_support_t m_ssl_support;
+    boost::asio::ssl::stream<boost::asio::ip::tcp::socket> socket_;
 
 		std::atomic<long> &m_ref_sock_count; // reference to external counter of existing sockets that we will ++/--
 	public:
 		// first counter is the ++/-- count of current sockets, the other socket_number is only-increasing ++ number generator
-		connection_basic(boost::asio::io_service& io_service, std::atomic<long> &ref_sock_count, std::atomic<long> &sock_number);
+		connection_basic(boost::asio::io_service& io_service, std::atomic<long> &ref_sock_count, std::atomic<long> &sock_number, ssl_support_t ssl, ssl_context_t &ssl_context);
 
 		virtual ~connection_basic() noexcept(false);
 
+		boost::asio::ip::tcp::socket& socket() { return socket_.next_layer(); }
+		ssl_support_t get_ssl_support() const { return m_ssl_support; }
+		void disable_ssl() { m_ssl_support = e_ssl_support_disabled; }
+
+		bool handshake(boost::asio::ssl::stream_base::handshake_type type)
+		{
+			return ssl_handshake(socket_, type, m_ssl_context);
+		}
+
+		template<typename MutableBufferSequence, typename ReadHandler>
+		void async_read_some(const MutableBufferSequence &buffers, ReadHandler &&handler)
+		{
+			if (m_ssl_support == e_ssl_support_enabled)
+				socket_.async_read_some(buffers, handler);
```

Comment:
Over my head.
Do you mean:
> socket_.async_read_some(buffers, handler);
should be:
> socket_.async_read_some(buffers, std::forward\<ReadHandler\>(handler));

If so I have no idea why but I'll do that once comfirmed.

Reply:
Yes, the handler should be "forwarded" just like in the snippet you gave in both cases here and in the write functions. Without this, a temporary `handler` (r-value) is copied instead of moved into the internally allocated asio object. The compiler doesn't automatically forward the object onto the next function; the programmer has the option when/where to pass-by-copy. Search for "universal references".

The difference in this exact situation will be that it saves one `shared_ptr` copy which has an internal atomic reference +/-.